### ISSUE TITLE
Need to add CFG Scale and CFG and context free grammars

### ISSUE THREAD

**Issue by ArEnSc** [undreamai/LLMUnity#30]

**Comment by amakropoulos** [comment]
Thanks for the issue!
I think you mean adding the following options from llama.cpp, is it correct?
--grammar / --grammar-file
--cfg-negative-prompt / --cfg-negative-prompt-file
--cfg-scale

If yes, sure I can do that.
I'm trying to think of a good way to add different options from llama.cpp without blowing up the UI.
Maybe I would separate some options that are probably less frequently in a usual chat setting.
and provide a field that the used can add arguments to the llama.cpp server command.
What do you think?

**Comment by ArEnSc** [comment]
yeah I wonder if this is supported onthe llamafile server, I checked I don't see those last 3 options

**Comment by amakropoulos** [comment]
I'm upgrading to llamafile 0.6, that contains those :)

**Comment by amakropoulos** [comment]
I have released v1.0.5 that allows grammars along with other parameters.
Unfortunately I can't set the negative-prompt and the scale because they are not exposed as arguments from the llama.cpp server :/

**Comment by ArEnSc** [comment]
> I have released v1.0.5 that allows grammars along with other parameters. Unfortunately I can't set the negative-prompt and the scale because they are not exposed as arguments from the llama.cpp server :/

yeah it was confusing since I saw args code that implied that it was availible from llamafile, I think ill or someone will need to go in there and make the changes to the server and understand how the CFG scale works once my project is off the ground, amazing work though btw! =) 
from my understanding CFG scale requires that we run inference twice and the server did drift away from the original PR made and it was never merged

**Comment by amakropoulos** [comment]
yes exactly, I also saw the llamafile args and confused it 🙈

**Comment by amakropoulos** [comment]
Old topic, but I revisited it :).
The llama.cpp server does not support CFG at the moment, only the main example does.
There is a WIP PR here: https://github.com/ggerganov/llama.cpp/pull/2217
